### ISSUE TITLE
Unable to query date field in Hive external table

### ISSUE THREAD

**Issue by ShabbySam** [elastic/elasticsearch-hadoop#2065]
### What kind an issue is this?

 Bug report.

### Issue description

when I query date field in Hive external table,I am getting below error `Error:Java.io.IOException:org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.ClassCastException:org.apache.hadoop.hive.serde2 .io.Timestampwritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampwritableV2 (state=,code=0)`

### Steps to reproduce

The mapping about data in es like below,Data already exists.
```
"data": {
    "mappings": {
      "_type": {
        "dynamic": "false",
        "numeric_detection": false,
        "properties": {
          "agent_id": {
            "type": "keyword"
          },
          "dt_server_time": {
            "type": "timestamp"
          }
        }
      }
    }
  }
```

I created an Hive external table to query the existing data of es like below

```
CREATE EXTERNAL TABLE ods_es_data_inc
(`agent_id`          STRING,
 `dt_server_time`    TIMESTAMP
) COMMENT 'bb_i_app'
STORED BY 'org.elasticsearch.hadoop.hive.EsStorageHandler'
TBLPROPERTIES(
    'es.resource'='data*',
    'es.nodes'='ip',
    'es.port'='port',
    'es.net.http.auth.user'='user',
    'es.net.http.auth.pass'='pass'
)
```
When I query data for external table like select count(*) from ods_es_data_inc I am getting below error.
`Error:Java.io.IOException:org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.ClassCastException:org.apache.hadoop.hive.serde2 .io.Timestampwritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampwritableV2 (state=,code=0)`

By searching for issues([#744](https://github.com/elastic/elasticsearch-hadoop/issues/744)), I think it may be the problem that es-hadoop does not match hive3. x.

Then,I try to use elasticsearch-hadoop[v7.13.0](https://github.com/elastic/elasticsearch-hadoop/labels/v7.13.0).But the problem remains.

Strack trace:

```
Because my test environment is an intranet, I cannot directly Strack trace information locally.You can tell me what you need, and then I'll write it down.
``` 

### Version Info

OS:         :  
JVM         :  
Hadoop/Spark:  
Hive : 3.1.0
ES-Hadoop   :  I have used 6.7.1 / 6.8.7 / 6.8.23 / 7.13.0 ;The problem remains.
ES          :  6.7.1

**Comment by masseyke** [comment]
A stack trace would definitely be helpful, but can you see if there is a log message like `Could not load org.apache.hadoop.hive.serde2.io.TimestampWritableV2. Continuing with legacy Timestamp class.` coming from [the code that is trying to load that class](https://github.com/elastic/elasticsearch-hadoop/blob/main/hive/src/main/java/org/elasticsearch/hadoop/hive/HiveValueReader.java#L53)? You definitely need es-hadoop 7.13.0 or higher. I'm wondering if you maybe just have a classpath problem?

**Comment by ShabbySam** [comment]
The Strack trace about the error like below.I hope this is useful information.
```
Error:java.io.IOException:org.apache.hadoop.hive.ql.metadata.HiveException:java.lang.ClassCastException:
org.apache.hadoop.hive.serde2.io.TimestampWritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampWritableV2(state=,code=0)
org.apache.hive.service.cli.HiveSQLException:java.io.IOException:org.apache.hadoop.hive.ql.metadata.HiveException:
java.lang.ClassCastException:org.apache.hadoop.hive.serde2.io.TimestampWritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampWritableV2
    at org.apache.hive.jdbc.Utils.verifySuccess(Utils.java:305)
    at org.apache.hive.jdbc.Utils.verifySuccessWithInfo(Utils.java:291)
    at org.apache.hive.jdbc.HiveQueryResultSet.next(HiveQueryResultSet.java:369)
    at org.apache.hive.beeline.BufferedRows.<init>(BufferedRows.java:56)
    at org.apache.hive.beeline.IncrementalRowsWithNormalization.<init>(IncrementalRowsWithNormalization,java:50)
    at org.apache.hive.beeline.BeeLine.print (BeeLine.java:2354)
    at org.apache.hive.beeline.Commands.executeInternal(Commands.java:1026)
    at org.apache.hive.beeline.Commands.execute(Commands.java:1201)
    at org.apache.hive.beeline.Commands.sql(Commands.java:1130)
    at org.apache.hive.beeline.BeeLine.dispatch(BeeLine.java:1529)
    at org.apache.hive.beeline.BeeLine.execute(BeeLine.java:1291)
    at org.apache.hive.beeline.BeeLine.begin(BeeLine.java:1072)
    at org.apache.hive.beeline.BeeLine.mainWithInputRedirection(BeeLine.java:541)
    at org.apache.hive.beeline.BeeLine.main(BeeLine.java:523)
Caused by:org.apache.hive.service.cli.HiveSQLException:java.io.IOException:org.apache.hadoop.hive.ql.metadata.HiveException:
java.lang.ClassCastException:org.apache.hadoop.hive.serde2.io.TimestampWritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampWritablev2
    at org.apache.hive.service.cli.operation.SQLOperation.getNextRowSet(SQLOperation.java:527)
    at org.apache.hive.service.cli.operation.OperationManager.getOperationNextRowSet (OperationManager.java:321)
    at org.apache.hive.service.cli.session.HiveSessionImpl.fetchResults(HiveSessionImpl.java:1071)
    at sun.reflect.GeneratedMethodAccessor80.invoke (Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:78)
    at org.apache.hive.service.cli.session.HiveSessionProxy.access$000(HiveSessionProxy.java:36)
    at org.apache.hive.service.cli.session.HiveSessionProxys1.run (HiveSessionProxy.java:63)
    at java.security.AccessController.doprivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:422)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1729)
    at org.apache.hive.service.cli.session.HiveSessionProxy.invoke (HiveSessionProxy.java:59)
    at com.sun.proxy.$Proxy47.fetchResults(Unknown Source)
    at org.apache.hive.service.cli.CLIService.fetchResults(CLIService.java:591)
    at org.apache.hive.service.cli.thrift.ThriftCLIService.FetchResults(ThriftCLIService.java:1707)
    at org.apache.hive.service.rpc.thrift.TCLIServicesprocessorsFetchResults.getResult (TCLIService.java:1837)
    at org.apache.hive.service.rpc.thrift.TCLIServicesProcessorsFetchResults.getResult(TCLIService.java:1822)
    at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
    at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
    at org.apache.hadoop.hive.metastore.security.HadoopThriftAuthBridge$Server$TUGIAssumingProcessor.process(HadoopThriftAuthBridge.java:648)
    at org.apache.thrift.server.TThreadPoolServer$Workerprocess.run(TThreadPoolServer.java:286)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutorsWorker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
Caused by:java.io.IOException:org.apache.hadoop.hive.ql.metadata.HiveException:java.lang.ClassCastException:org.apache.hadoop.hive.serde2.io.TimestampWritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampWritableV2
    at org.apache.hadoop.hive.ql.exec.FetchTask.fetch(FetchTask.java:166)
    at org.apache.hadoop.hive.ql.Driver.getResults(Driver.java:3260)
    at org.apache.hadoop.hive.ql.reexec.ReExecDriver.getResults(ReExecDriver.java:252)
    at org.apache.hive.service.cli.operation.SQLOperation.getNextRowset (SQLOperation.java:522)
    ··24more
Caused by:org.apache.hadoop.hive.ql.metadata.HiveException:java.lang.ClassCastException:
org.apache.hadoop.hive.serde2.io.TimestampWritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampWritableV2
    at org.apache.hadoop.hive.ql.exec.ListSinkoperator.process (ListSinkOperator.java:98)
    at org.apache.hadoop.hive.ql.exec.Operator.baseForward(Operator.java:995)
    at org.apache.hadoop.hive.ql.exec.Operator.forward (Operator.java:941)
    at org.apache.hadoop.hive.ql.exec.Operator.forward(Operator.java:928)
    at org.apache.hadoop.hive.ql.exec.Limitoperator.process(Limitoperator.java:63)
    at org.apache.hadoop.hive.gl.exec.Operator.baseForward(Operator.java:995)
    at org.apache.hadoop.hive.ql.exec.Operator.forward (Operator.java:941)
    at org.apache.hadoop.hive.ql.exec.Operator.forward(Operator.java:928)
    at org.apache.hadoop.hive.ql.exec.Selectoperator.process(Selectoperator.java:95)
    at org.apache.hadoop.hive.ql.exec.Operator.baseForward(Operator.java:995)
    at org.apache.hadoop.hive.gl.exec.Operator.forward(Operator.java:941)
    at org.apache.hadoop.hive.gl.exec.TableScanOperator.process(TableScanOperator.java:125)
    at org.apache.hadoop.hive.ql.exec.FetchOperator.pushRow(FetchOperator.java:528)
    at org.apache.hadoop.hive.ql.exec.FetchOperator.pushRow(FetchOperator.java:520)
    at org.apache.hadoop.hive.ql.exec.FetchTask.fetch(FetchTask.java:150)
    ..27 more
Caused by:java.lang.ClassCastException:org.apache.hadoop.hive.serde2.io.TimestampWritable cannot be cast to org.apache.hadoop.hive.serde2.io.TimestampWritableV2
    at org.apache.hadoop.hive.serde2.objectinspector.primitive.WritableTimestampObjectInspector.getPrimitiveJavaobject(WritableTimestampObjectInspector.java:39)
    at org.apache.hadoop.hive.serde2.objectinspector.primitive.WritableTimestampObjectInspector.getPrimitiveJavaobject(WritableTimestampObjectInspector.java:25)
    at org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorUtils.copyToStandardObject(ObjectInspectorUtils.java:422)
    at org.apache.hadoop.hive.serde2.SerDeUtils.toThriftPayload(SerDeUtils.java:173)
    at org.apache.hadoop.hive.serde2.thrift.ThriftFormatter.convert(ThriftFormatter.java:49)
    at org.apache.hadoop.hive.ql.exec.ListSinkoperator.process(ListSinkoperator.java:94)
```

beeline has a flag --verbose=[true/false] that will output more information, I use that option to get back more information about stacktraces.

If you need the log in Yarn, I will try to write it down tomorrow.

**Comment by masseyke** [comment]
OK so your hive server is getting back a row with a TImestampWritable (so your hive server is running the expected newer version). I would guess it's because es-hadoop is getting the old hive jar in its classpath. It would be good to check the yarn logs for the `Could not load org.apache.hadoop.hive.serde2.io.TimestampWritableV2`... message, and to check the classpath you're running es-hadoop with.

**Comment by ShabbySam** [comment]
I have solved this problem, but I have no idea how to solve it.
Now let me talk about my operation process before the error disappears

**Steps to reproduce**
First of all,my query statement was too simple and did not involve MR engine operations, so there were no logs in the YARN web UI.Therefore, I plan to write some complex query statements involving the timestemp field.
At the same time, I remembered that switching Hive to local mode might solve some class loading issues. So I switched Hive to local mode and ran a query with a _where_ condition. 
```
select 
    agent_id,
    dt_server_time
from ods_es_data_inc
where agent_id=123;
```
It ran successfully without any errors and the results were correct. I switched Hive back to cluster mode and ran the query statement again, which also succeeded without any problems. 

Finally, I wrote a simple query statement that did not require MR operations to run, and it also ran successfully without any bugs.
```
select 
    agent_id,
    dt_server_time
from ods_es_data_inc
limit 1;
```
